File: application/app/src/main/java/com/yahyeet/boardbook/presenter/SearchAdapter.java
package com.yahyeet.boardbook.presenter;

import android.widget.Filter;

import androidx.recyclerview.widget.RecyclerView;

import com.yahyeet.boardbook.model.entity.AbstractEntity;

import java.util.ArrayList;
import java.util.List;

public abstract class SearchAdapter<T extends AbstractEntity> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private List<T> database;
	private List<T> allEntities;
	private Filter filter;

	public SearchAdapter(List<T> database) {

		if(database != null)
			this.database = database;
		else
			this.database = new ArrayList<>();


		allEntities = new ArrayList<>(this.database);
		filter = createNewFilter();

	}

	@Override
	public int getItemCount() {
		return database.size();
	}

	public Filter getFilter() {
		return filter;
	}

	public List<T> getDatabase() {
		return database;
	}

	public List<T> getAllEntities() {
		if(database.size() > 0 && allEntities.size() == 0)
			allEntities = new ArrayList<>(database);
		return allEntities;
	}

	protected abstract Filter createNewFilter();
}

File: application/app/src/main/java/com/yahyeet/boardbook/presenter/SearchAdapter.kt
package com.yahyeet.boardbook.presenter

import android.widget.Filter

import androidx.recyclerview.widget.RecyclerView

import com.yahyeet.boardbook.model.entity.AbstractEntity

import java.util.ArrayList

abstract class SearchAdapter<T : AbstractEntity>(database: List<T>?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var database: List<T>? = null
        private set
    private var allEntities: List<T>? = null
    val filter: Filter

    init {

        if (database != null)
            this.database = database
        else
            this.database = ArrayList()


        allEntities = ArrayList(this.database!!)
        filter = createNewFilter()

    }

    override fun getItemCount(): Int {
        return database!!.size
    }

    fun getAllEntities(): List<T> {
        if (database!!.size > 0 && allEntities!!.size == 0)
            allEntities = ArrayList(database!!)
        return allEntities
    }

    protected abstract fun createNewFilter(): Filter
}
